File: Sources/ShinyUI/Runtime/Existential.swift
//
//  File.swift
//  
//
//  Created by Angel Landoni on 16/2/21.
//

import Runtime

extension Existential {
    @_optimize(none)
    public static func fromProtocol<T>(_ value: T) -> Existential {
        return withUnsafePointer(to: value) { pointer in
            map_to_existential(pointer)
        }
    }
}


extension Existential: CustomStringConvertible {
    public var description: String {
        """
        [!] Existential:
            [+] ValueA: \(valueA)
            [+] ValueB: \(valueB)
            [+] ValueC: \(valueC)
            [+] Metadata: \n\t\(nest: commondMetadata.description)
            [*] Witness table: \(witnessTablesPtr)
        """
    }
}

extension Optional: CustomStringConvertible {
    public var description: String {
        if let data = self {
            return "\(data)"
        }
        return "NIL"
    }
}

File: Sources/Runtime/Existential.c
//
//  Existential.c
//  
//
//  Created by Angel Landoni on 15/2/21.
//

#include "include/Existential.h"

struct Existential map_to_existential(void const *existential) {
    const unsigned long wordSize = sizeof(WORD);
    struct Existential out;
    
    unsigned long offset = 0;
    
    // Copy value buffers into the struct.
    memcpy(&out, existential, wordSize * NUMBER_OF_BUFFERS_IN_EXISTENTIAL);
    offset += wordSize * NUMBER_OF_BUFFERS_IN_EXISTENTIAL;
    
    // Copy metadata.
    void *metadataAddress = NULL;
    memcpy(&metadataAddress, existential + offset, wordSize);
    offset += wordSize;
    
    struct Metadata metadata = map_to_metadata(metadataAddress);
    
    out.commondMetadata = metadata;
    
    // Copy witness table.
    void *witnessTablesPtr = NULL;
    memcpy(&witnessTablesPtr, existential + offset, wordSize);
    
    out.witnessTablesPtr = witnessTablesPtr;
    
    return out;
}
